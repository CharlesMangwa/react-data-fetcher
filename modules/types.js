/* @flow */
/* eslint-disable no-use-before-define */

import PropTypes from 'prop-types'

// FLOW

export type Interceptor = (InterceptedData) => ?RequestToApi;
export type Context = {
  rdfApi: string,
  rdfHeaders: Object,
  rdfInterceptor: ?Interceptor,
  rdfLoader: React$Node,
  rdfStore: Object,
  rdfTimeout: number,
}

export type ErrorContent = {
  request: XMLHttpRequest,
  response: String | Object,
}

export type Error = {
  content: ErrorContent,
  message: 'Something went wrong during the request',
  url?: string,
}

export type Method =
  'DELETE'
| 'FORM_DATA'
| 'GET'
| 'HEAD'
| 'PATCH'
| 'POST'
| 'PUT'
| 'TRACE'

export type DefaultProps = {
  body: Object,
  method: Method,
  params: Object,
}

export type Progress = {
  bubbles: boolean,
  cancelable: boolean,
  lengthComputable: boolean,
  loaded: number,
  target: EventTarget,
  total: number,
  type: string,
}

export type ReturnedData = {
  data?: Object,
  error?: Error,
  isOK?: boolean,
  request?: XMLHttpRequest,
  status?: number,
  store?: Object,
}

export type Props = {
  body?: Object,
  children?: React$StatelessFunctionalComponent<?ReturnedData | Error>,
  component?: React$ComponentType<?ReturnedData | Error>,
  headers?: Object,
  loader?: React$Node,
  method: Method,
  onError?: (?ReturnedData | Error) => void,
  onFetch?: (?ReturnedData | Error) => void,
  onLoad?: Function,
  onProgress?: (Progress) => void,
  onTimeout?: Function,
  onIntercept?: Interceptor,
  params?: Object,
  path?: string,
  refetchKey?: any,
  render?: React$StatelessFunctionalComponent<?ReturnedData | Error>,
  resultOnly?: boolean,
  timeout?: number,
  url?: string,
}

export type RequestToApi = {
  body?: Object,
  headers?: Object,
  method: Method,
  onProgress?: (Progress) => void,
  onTimeout?: Function,
  onIntercept?: ?Interceptor,
  params?: Object,
  url: string,
  timeout?: number,
}

export type InterceptedData = {
  currentParams: RequestToApi,
  request: XMLHttpRequest,
  status: number
}

export type Store = {
  subscribe: Function,
  dispatch: Function,
  getState: () => Object,
}

export type ProviderProps = {
  api: string,
  children: React$Node,
  headers?: Object,
  loader?: React$Node,
  onIntercept?: Interceptor,
  store?: Store,
  timeout?: number,
}

// PROPTYPES
export const methodShape = PropTypes.oneOf([
  'DELETE',
  'FORM_DATA',
  'GET',
  'HEAD',
  'PATCH',
  'POST',
  'PUT',
  'TRACE',
])

export const storeShape = PropTypes.shape({
  getState: PropTypes.func,
})
