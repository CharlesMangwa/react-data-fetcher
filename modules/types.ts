// import PropTypes from "prop-types";

import React from "react";

export type Interceptor = (interceptedData: IInterceptedData) => IRequestToApi | void;

export interface IContext {
  api?: string;
  headers?: {[s: string]: string};
  onIntercept?: Interceptor;
  loader?: React.ReactNode;
  store?: object;
  timeout?: number;
}

export interface IErrorContent {
  request: XMLHttpRequest;
  response: string | object;
}

export interface IError {
  content: IErrorContent;
  message: "Something went wrong during the request";
  url?: string;
}

export type Method =
  | "DELETE"
  | "FORM_DATA"
  | "GET"
  | "HEAD"
  | "PATCH"
  | "POST"
  | "PUT"
  | "TRACE";

export interface IReturnedData {
  data?: object;
  error?: IError;
  isOK?: boolean;
  request?: XMLHttpRequest;
  status?: number;
  store?: object;
}

export interface IDataFetcher {
  fetchData(): Promise<void>;
}

export interface IProps {
  body: {[s: string]: any};
  cancel?: boolean;
  children?: React.ReactNode;
  component?: React.ComponentType<IReturnedData | IError>;
  context: IContext;
  headers?: {[s: string]: string};
  loader?: React.ReactNode;
  method: Method;
  onError?: (data?: IReturnedData | IError) => void;
  onFetch?: (data?: IReturnedData | IError) => void;
  onIntercept?: Interceptor;
  onLoad?: () => void;
  onProgress?: (progressEvent: ProgressEvent) => void;
  onTimeout?: () => void;
  params?: {[s: string]: any};
  path?: string;
  refetchKey?: any;
  render?: React.SFC<IReturnedData | IError>;
  resultOnly?: boolean;
  timeout: number;
  url?: string;
}

export interface IRequestToApi {
  body?: {[s: string]: any};
  cancel?: boolean;
  headers?: {[s: string]: string};
  method: Method;
  onProgress?: (progressEvent: ProgressEvent) => void;
  onIntercept?: Interceptor;
  onTimeout?: () => void;
  params?: {[s: string]: any};
  url: string;
  timeout?: number;
}

export interface IInterceptedData {
  currentParams: IRequestToApi;
  request: XMLHttpRequest;
  status: number;
}

export interface IProviderProps {
  api: string;
  children: React.ReactNode;
  headers?: {[s: string]: string};
  loader?: React.ReactNode;
  onIntercept?: Interceptor;
  store?: object;
  timeout?: number;
}

/*
export const contextShape = PropTypes.shape({
  api: PropTypes.string,
  headers: PropTypes.object,
  loader: PropTypes.element,
  onIntercept: PropTypes.func,
  store: PropTypes.object,
  timeout: PropTypes.number,
});

export const methodShape = PropTypes.oneOf([
  "DELETE",
  "FORM_DATA",
  "GET",
  "HEAD",
  "PATCH",
  "POST",
  "PUT",
  "TRACE",
]);
*/
