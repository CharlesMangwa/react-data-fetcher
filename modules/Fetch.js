/* @flow */

import { Children, Component, createElement } from 'react'
import PropTypes from 'prop-types'
import invariant from 'invariant'

import requestToApi from './requestToApi'
import {
  type Context,
  type ErrorContent,
  type Error,
  type Props,
  type ReturnedData,
  methodShape,
} from './types'

const isEmptyChildren = children => Children.count(children) === 0

class Fetch extends Component<Props> {
  _data: ?ReturnedData | Error = undefined
  _didCallOnLoad = false
  _isLoaded = false
  _isUnmounted = false

  static propTypes = {
    body: PropTypes.object,
    children: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
    component: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
    method: methodShape,
    loader: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
    onError: PropTypes.func,
    onFetch: PropTypes.func,
    onLoad: PropTypes.func,
    onProgress: PropTypes.func,
    onTimeout: PropTypes.func,
    params: PropTypes.object,
    path: PropTypes.string,
    refetchKey: PropTypes.any,
    render: PropTypes.func,
    resultOnly: PropTypes.bool,
    url: PropTypes.string,
    timeout: PropTypes.number,
  }

  static contextTypes = {
    rdfApi: PropTypes.string,
    rdfHeaders: PropTypes.object,
    rdfInterceptor: PropTypes.func,
    rdfLoader: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
    rdfStore: PropTypes.object,
    rdfTimeout: PropTypes.number,
  }

  static defaultProps = {
    body: {},
    children: undefined,
    component: undefined,
    loader: undefined,
    method: 'GET',
    onError: undefined,
    onFetch: undefined,
    onLoad: undefined,
    onProgress: undefined,
    onTimeout: undefined,
    params: {},
    path: undefined,
    refetchKey: false,
    render: undefined,
    resultOnly: false,
    url: undefined,
    timeout: -1,
  }

  componentWillMount() {
    this._validateProps(this.props, this.context)
    if (this.props.onLoad && !this._didCallOnLoad) {
      this._didCallOnLoad = true
      this.props.onLoad()
    }
  }

  componentDidMount() {
    if (this.props.path === 'store') {
      this._handleData({
        data: this.context.rdfStore,
        isOK: true,
      })
    }
    else this._fetchData(this.props, this.context)
  }

  componentWillReceiveProps(nextProps: Props, nextContext: Context) {
    const { onLoad, path, refetchKey } = this.props

    this._validateProps(nextProps, nextContext)

    if (onLoad && !this._didCallOnLoad) {
      this._didCallOnLoad = true
      onLoad()
    }

    if (path === 'store') {
      this._handleData({
        data: this.context.rdfStore,
        isOK: true,
      })
    }
    else if (
      nextProps.path !== path ||
      nextProps.refetchKey !== refetchKey
    )
      this._fetchData(nextProps, nextContext)
  }

  componentWillUnmount() {
    this._isUnmounted = true
  }

  shouldComponentUpdate(nextProps: Props) {
    if (this.props.children !== nextProps.children) return true
    if (this.props.loader !== nextProps.loader) return true
    if (this.props.onError !== nextProps.onError) return true
    if (this.props.onFetch !== nextProps.onFetch) return true
    if (this.props.onLoad !== nextProps.onLoad) return true
    if (this.props.path !== nextProps.path) return true
    if (this.props.params !== nextProps.params) return true
    if (this.props.refetchKey !== nextProps.refetchKey) return true
    if (this.props.render !== nextProps.render) return true
    if (this._isLoaded) return true
    if (this._data) return true
    return false
  }

  _fetchData = async (props: Props, context: Context): Promise<void> => {
    const {
      body,
      headers,
      method,
      onIntercept,
      onProgress,
      onTimeout,
      params,
      path,
      url,
      timeout,
    } = props
    let route: ?string
    let timeoutValue: number = 0

    if (path) route = `${context.rdfApi || ''}${path}`
    else route = url

    if (context.rdfTimeout && timeout === -1)
      timeoutValue = context.rdfTimeout
    else if (!context.rdfTimeout && timeout)
      timeoutValue = Math.max(0, timeout)
    else if (context.rdfTimeout && timeout) {
      timeoutValue = timeout === -1
        ? context.rdfTimeout
        : timeout
    }

    try {
      const apiResponse = await requestToApi({
        url: route || '',
        body: { ...body },
        headers: { ...context.rdfHeaders, ...headers },
        method,
        onTimeout,
        onProgress,
        onIntercept: onIntercept || context.rdfInterceptor,
        params: { ...params },
        timeout: timeoutValue,
      })
      if (!this._isUnmounted) {
        this._handleData({
          ...apiResponse,
          store: context.rdfStore,
        })
      }
    }
    catch (error) {
      if (!this._isUnmounted) {
        this._handleData({
          error: {
            content: error,
            message: 'Something went wrong during the request',
            url: route,
          },
          isOK: false,
          store: context.rdfStore,
        })
        if (process.env.NODE_ENV !== 'production') {
          invariant(
            !error,
            `<Fetch> tried to call the route "${String(route)}" ` +
              `with "${String(method).toUpperCase()}" method ` +
              'but resolved with the following error: %s\n',
            this._printError(error),
          )
        }
      }
    }
  }

  _handleData = (result: ReturnedData): void => {
    if (!this._isUnmounted) {
      this._isLoaded = true
      this.props.resultOnly
        ? (this._data = result.error || result.data)
        : (this._data = result)
      this._returnData(result)
    }
  }

  _printError = (error: ErrorContent): string => (
    error.response && JSON.stringify(error.response).length
      ? typeof error.response === 'string'
        ? error.response
        : typeof error.response === 'object'
          ? JSON.stringify(error.response)
          : `${error.response}. Sorry <Fetch> couldn't turned this into a readable string. `
            + 'Check error.content.request to see what happened.'
      : error.request._response
        ? typeof error.request._response === 'string'
          ? error.request._response
          : typeof error.request._response === 'object'
            ? JSON.stringify(error.request._response)
            : `${String(error.request._response)}. Sorry <Fetch> couldn't turned this into a readable string. `
              + 'Check error.content.request to see what happened.'
        : " .Sorry <Fetch> couldn't turned this into a readable string. "
            + 'Check error.content.request to see what happened.'
  )

  _renderLoader = (): React$Node => {
    const { rdfLoader } = this.context
    const { loader } = this.props

    if (rdfLoader && !loader) {
      return typeof rdfLoader === 'function'
        ? rdfLoader() : rdfLoader
    }
    else if (!rdfLoader && loader) {
      return typeof loader === 'function'
        ? loader() : loader
    }
    else if (rdfLoader && loader) {
      return typeof loader === 'function'
        ? loader() : loader
    }

    return null
  }

  _returnData = (result: ReturnedData): void => {
    const { onError, onFetch } = this.props

    if (onFetch) onFetch(this._data)

    if (result.error && onError) onError(this._data)

    if (!this._isUnmounted) this.forceUpdate()
  }

  _validateProps = (props: Props, context: Context): void => {
    const { rdfApi, rdfStore, rdfTimeout } = context
    const {
      children,
      component,
      onTimeout,
      onFetch,
      path,
      render,
      timeout,
      url,
    } = props

    invariant(
      path || url,
      'You must provide a `url` or a `path` to <Fetch>',
    )

    if (path) {
      invariant(
        path && rdfApi,
        'You must implement <ConnectedFetch> at the root of your '
          + 'app and provide an `api` in order to use `path`',
      )
    }

    if (path === 'store') {
      invariant(
        path && rdfStore,
        'You must implement <ConnectedFetch> at the root of your '
          + 'app and provide a `store` in order to use `path="store"`',
      )
    }

    if (onTimeout) {
      invariant(
        (typeof timeout === 'number' && timeout >= 0) ||
        (typeof rdfTimeout === 'number' && rdfTimeout >= 0),
        'You must provide a `timeout` number in ms to <Fetch> or <ConnectedFetch> '
          + 'in order to use `onTimeout`',
      )
    }

    invariant(
      children || component || render || onFetch,
      'You must provide at least one of the following '
        + 'to <Fetch>: children, `component`, `onFetch`, `render`',
    )
  }

  render(): React$Node {
    const { children, component, render } = this.props

    if (!this._isLoaded && !this._isUnmounted)
      return this._renderLoader()

    if (this._isLoaded && !this._isUnmounted) {
      if (component) return createElement(component, this._data)

      if (typeof render === 'function') return render(this._data)

      if (typeof children === 'function') return children(this._data)

      if (children && !isEmptyChildren(children))
        return Children.only(children)
    }

    return null
  }
}

export default Fetch
