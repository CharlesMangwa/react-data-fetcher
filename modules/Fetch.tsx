import invariant from "invariant";
import React, { Children, Component, createElement } from "react";
import { polyfill } from "react-lifecycles-compat";
// import PropTypes from "prop-types";

import { Consumer } from "./FetchProvider";
import requestToApi from "./requestToApi";
import {
  IDataFetcher,
  IError,
  IErrorContent,
  IProps,
  IReturnedData,
  Method,
} from "./types";
import { isEmptyChildren } from "./util";

const defaultProps = {
  body: {},
  cancel: false,
  children: undefined,
  component: undefined,
  context: {},
  loader: undefined,
  method: "GET" as Method,
  onError: undefined,
  onFetch: undefined,
  onLoad: undefined,
  onProgress: undefined,
  onTimeout: undefined,
  params: {},
  path: undefined,
  refetchKey: undefined,
  render: undefined,
  resultOnly: false,
  timeout: -1,
  url: undefined,
};

const defaultize = (props: Partial<IProps>): IProps => ({
  ...defaultProps,
  ...props,
});

class Fetch extends Component<Partial<IProps>> implements IDataFetcher {

  private data?: IReturnedData | IError;
  private didCallOnLoad = false;
  private isLoaded = false;
  private isUnmounted = false;

  public componentDidMount() {
    const { context, onLoad, path } = defaultize(this.props);
    if (onLoad && !this.didCallOnLoad) {
      this.didCallOnLoad = true;
      onLoad();
    }
    if (path === "store") {
      this.handleData({
        data: context.store,
        isOK: true,
      });
    } else {
      this.fetchData();
    }
  }

  public componentDidUpdate(prevProps: Partial<IProps>) {
    this.validateProps(this.props);
    const { context, onLoad, path, refetchKey } = defaultize(this.props);
    if (onLoad && !this.didCallOnLoad) {
      this.didCallOnLoad = true;
      onLoad();
    }

    if (refetchKey !== prevProps.refetchKey) {
      if (path === "store") {
        this.isLoaded = true;
        this.handleData({
          data: context.store,
          isOK: true,
        });
      } else {
        this.isLoaded = false;
        this.fetchData();
      }
    }
  }

  public componentWillUnmount() {
    this.isUnmounted = true;
  }

  public shouldComponentUpdate(nextProps: Partial<IProps>) {
    const defaultized = defaultize(this.props);
    const nextDefaultized = defaultize(nextProps);
    if (defaultized.cancel !== nextDefaultized.cancel) { return true; }
    if (defaultized.children !== nextDefaultized.children) { return true; }
    if (defaultized.loader !== nextDefaultized.loader) { return true; }
    if (defaultized.onError !== nextDefaultized.onError) { return true; }
    if (defaultized.onFetch !== nextDefaultized.onFetch) { return true; }
    if (defaultized.onLoad !== nextDefaultized.onLoad) { return true; }
    if (defaultized.path !== nextDefaultized.path) { return true; }
    if (defaultized.params !== nextDefaultized.params) { return true; }
    if (defaultized.refetchKey !== nextDefaultized.refetchKey) { return true; }
    if (defaultized.render !== nextDefaultized.render) { return true; }
    if (this.isLoaded) { return true; }
    if (this.data) { return true; }
    return false;
  }

  public render() {
    const { children, component, render } = defaultize(this.props);

    if (!this.isLoaded && !this.isUnmounted) { return this.renderLoader(); }

    if (this.isLoaded && !this.isUnmounted) {
      if (component) {
        return createElement(component, this.data);
      }
      if (typeof render === "function" && this.data) {
        return render(this.data);
      }
      if (typeof children === "function" && this.data) {
        return children(this.data);
      }
      if (children && !isEmptyChildren(children)) {
        return Children.only(children);
      }
    }

    return null;
  }

  public async fetchData(): Promise<void> {
    const {
      body,
      cancel,
      context,
      headers,
      method,
      onIntercept,
      onProgress,
      onTimeout,
      params,
      path,
      url,
      timeout,
    } = defaultize(this.props);
    const route = path ? `${context.api || ""}${path}`  : url;
    const timeoutValue = context.timeout && timeout === -1 ?
      context.timeout :
        !context.timeout && timeout ?
        Math.max(0, timeout) :
          context.timeout && timeout ?
          timeout === -1 ? context.timeout : timeout :
            0;

    try {
      const apiResponse = await requestToApi({
        body: { ...body },
        cancel,
        headers: { ...context.headers, ...headers },
        method,
        onIntercept: onIntercept || context.onIntercept,
        onProgress,
        onTimeout,
        params: { ...params },
        timeout: timeoutValue,
        url: route || "",
      });
      if (!this.isUnmounted) {
        this.handleData({
          ...apiResponse,
          store: context.store,
        });
      }
    } catch (error) {
      if (!this.isUnmounted) {
        this.handleData({
          error: {
            content: error,
            message: "Something went wrong during the request",
            url: route,
          },
          isOK: false,
          store: context.store,
        });
        if (process.env.NODE_ENV !== "production") {
          invariant(
            !error,
            `<Fetch /> tried to call the route "${String(route)}" ` +
              `with "${String(method).toUpperCase()}" method ` +
              "but resolved with the following error: %s\n",
            this.printError(error),
          );
        }
      }
    }
  }

  private handleData(result: IReturnedData) {
    const { resultOnly } = defaultize(this.props);
    if (!this.isUnmounted) {
      this.isLoaded = true;
      resultOnly
        ? (this.data = result.error || result.data)
        : (this.data = result);
      this.returnData(result);
    }
  }

  private printError(error: IErrorContent): string {
    return error.response && JSON.stringify(error.response).length
      ? typeof error.response === "string"
        ? error.response
        : typeof error.response === "object"
          ? JSON.stringify(error.response, null, 2)
          : `${error.response}. "Check error.content to see what happened.`
      : "Check error.content to see what happened.";
  }

  private renderLoader(): React.ReactNode {
    const { context, loader } = defaultize(this.props);

    if (context.loader && !loader) {
      return typeof context.loader === "function"
        ? context.loader()
        : context.loader;
    }
    if (!context.loader && loader) {
      return typeof loader === "function" ? loader() : loader;
    }
    if (context.loader && loader) {
      return typeof loader === "function" ? loader() : loader;
    }

    return null;
  }

  private returnData(result: IReturnedData) {
    const { onError, onFetch } = defaultize(this.props);

    if (onFetch) { onFetch(this.data); }

    if (result.error && onError) { onError(this.data); }

    if (!this.isUnmounted) { this.forceUpdate(); }
  }

  private validateProps(props: Partial<IProps>) {
    const {
      children,
      component,
      context,
      onTimeout,
      onFetch,
      path,
      render,
      timeout,
      url,
    } = defaultize(props);

    invariant(path || url, "You must provide a `url` or a `path` to <Fetch />");

    if (path) {
      invariant(
        path && context.api,
        "You must implement <FetchProvider> at the root of your " +
          "app and provide a URL to `value.api` prop in order to use `path`",
      );
    }

    if (path === "store") {
      invariant(
        path && context.store,
        "You must implement <FetchProvider> at the root of your " +
          "app and provide an object to `value.store` prop " +
          "in order to use `store`",
      );
    }

    if (onTimeout) {
      invariant(
        (typeof timeout === "number" && timeout >= 0) ||
          (typeof context.timeout === "number" && context.timeout >= 0),
        "You must provide a `timeout` number in ms to <Fetch /> or " +
          "<FetchProvider> in order to use `onTimeout`",
      );
    }

    invariant(
      children || component || render || onFetch,
      "You must provide at least one of the following " +
        "to <Fetch />: children, `component`, `onFetch`, `render`",
    );
  }
}

const withContext = <P extends object>(
  FetchComponent: React.ComponentType<P>,
): React.FC<P & Partial<IProps>> => (props: Partial<IProps>) => (
  <Consumer>{
    (data) => <FetchComponent {...props as P} context={data} />
  }</Consumer>
);

export default withContext(polyfill(Fetch));
export { Fetch };
