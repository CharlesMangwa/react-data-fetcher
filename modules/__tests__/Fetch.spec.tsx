import React, { Component } from "react";
import TestRenderer from "react-test-renderer";
import { createRenderer, ShallowRenderer } from "react-test-renderer/shallow";

import { Fetch as TestFetch } from "../Fetch";
import { Fetch, FetchProvider } from "../index";
import { IDataFetcher, IError, IReturnedData } from "../types";

describe("A <Fetch />", () => {
  let fn: jest.Mock;
  let renderer: ShallowRenderer;

  beforeEach(() => {
    fn = jest.fn();
    renderer = createRenderer();
  });

  afterEach(() => jest.clearAllMocks());

  it("should render component `children` correctly", () => {
    expect.assertions(1);
    const component = TestRenderer.create(
      <Fetch url="https://api.github.com/users/octocat">
        <div />
      </Fetch>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render component `children` correctly", () => {
    expect.assertions(1);
    const component = TestRenderer.create(
      <Fetch url="https://api.github.com/users/octocat">
        <div />
      </Fetch>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render & call function children correctly", () => {
    expect.assertions(1);
    const component = TestRenderer.create(
      <Fetch url="https://api.github.com/users/octocat">
        {() => fn() || null}
      </Fetch>,
    );

    const instance = component.root;
    instance.props.children();

    expect(fn).toHaveBeenCalled();
  });

  it("should re-render only when necessary", () => {
    expect.assertions(4);
    const render = () => null;
    const params = { page: 1 };

    const component = TestRenderer.create(
      <TestFetch
        url="https://api.github.com/users/octocat"
        cancel={false}
        loader={render}
        onError={fn}
        onFetch={fn}
        onLoad={fn}
        path={undefined}
        params={params}
        refetchKey={false}
        render={undefined}
      >
        {render}
      </TestFetch>,
    );

    const instance = component.getInstance();
    if (instance) {
      const spy1 = jest.spyOn(instance as unknown as IDataFetcher, "fetchData");
      const spy2 = jest.spyOn(instance as unknown as Component, "shouldComponentUpdate");

      component.update(
        <TestFetch
          url="https://api.github.com/users/octocat"
          cancel={false}
          loader={render}
          onError={fn}
          onFetch={fn}
          onLoad={fn}
          path={undefined}
          params={params}
          refetchKey={false}
          render={undefined}
        >
          {render}
        </TestFetch>,
      );

      expect(spy1).not.toHaveBeenCalled();
      expect(spy2).toHaveReturnedWith(false);

      component.update(
        <TestFetch url="https://api.github.com/users/octocat" refetchKey>
          <div />
        </TestFetch>,
      );

      expect(spy1).toHaveBeenCalled();
      expect(spy2).toHaveReturnedWith(true);
    }
  });

  it("should call `onLoad` when passed", () => {
    expect.assertions(1);
    renderer.render(
      <TestFetch
        onLoad={fn}
        url="https://api.github.com/users/octocat"
        render={() => null}
      />,
    );

    expect(fn).toHaveBeenCalled();
  });

  it("should call `loader` when passed", () => {
    expect.assertions(1);
    renderer.render(
      <TestFetch
        loader={fn}
        url="https://api.github.com/users/octocat"
        render={() => null}
      />,
    );

    expect(fn).toHaveBeenCalled();
  });

  it("should propagate `store` correctly", () => {
    expect.assertions(1);
    const context = { api: "https://api.github.com", store: { cats: 42 } };
    const expectedData = {
      cats: 42,
    };
    let receivedData;

    TestRenderer.create(
      <FetchProvider value={context}>
        <Fetch
          resultOnly
          path="store"
          onFetch={(data) => (receivedData = data || null)}
        />
      </FetchProvider>,
    );

    expect(receivedData).toMatchObject(expectedData);
  });

  it("should throw when it is not rendered in the context of a <FetchProvider>", () => {
    expect.assertions(1);
    expect(() =>
      renderer.render(<TestFetch path="store">{() => null}</TestFetch>),
    ).toThrow();
  });

  it("should throw when no `url` nor `path` is passed", () => {
    expect.assertions(1);
    expect(() => renderer.render(<TestFetch>{() => null}</TestFetch>)).toThrow();
  });

  it("should throw when `onTimeout` is passed, but no `timeout`", () => {
    expect.assertions(1);
    expect(() =>
      renderer.render(
        <TestFetch
          url="https://api.github.com/users/octocat"
          onTimeout={() => fn()}
        >
          {() => null}
        </TestFetch>,
      ),
    ).toThrow();
  });

  it("should throw when no `children`, `component`, `onFetch`, `render` prop is passed", () => {
    expect.assertions(1);
    expect(() =>
      renderer.render(<TestFetch url="https://api.github.com/users/octocat" />),
    ).toThrow();
  });
});
