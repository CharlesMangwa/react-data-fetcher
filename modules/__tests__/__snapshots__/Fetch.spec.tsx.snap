// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <Fetch /> should render component \`children\` correctly 1`] = `null`;

exports[`A <Fetch /> should render component \`children\` correctly 2`] = `null`;
